src/math/math.cr

```diff
@@ -332,6 +332,28 @@ module Math
     sqrt(value.to_f)
   end
 
+  # Calculates the integer square root of *value*.
+  def isqrt(value : Int)
+    raise ArgumentError.new "Input must be non-negative integer" if value < 0
+    return value if value < 2
+
+    res = typeof(value).zero
+    bit = 1 << ((sizeof(typeof(value)) << 3) - 2)
```

Comment:
`typeof` doesn't work if the argument has a union type.
The former should be `value.class.zero`. I'm not sure about a proper replacement for `size_of`. I suppose we're missing a `Int.bit_length` method for that.

Reply:
https://github.com/crystal-lang/crystal/pull/8924.  Is already a thing.

Reply:
Yes, I was searching for exactly that, but could not spot anything. Another approach is a macro that unwraps to implementations for each `Int` type, but seems a little less... elegant.

Reply:
`Int#bit_length` returns the position of the highest bit. It's the bit length of the value, not the type.

Reply:
We could do something like this for the moment:

```crystal
struct Int
  # :nodoc:
  def bitsize
    sizeof(typeof(self)) << 3
  end
end
```

Reply:
Thoughts on be4321d?

Reply:
Wait, looking at the next expression which shifts back by `leading_zeros_count`, I would expect that we could actually use `Int#bit_length` here.

If nothing else, for positive values the type's bit length should be `leading_zeros_count + bit_length`, right?

Reply:
@straight-shoota looks like we hit a concurrency bug with our comms :)

Implementation referenced above uses less operations than calling `Int#bit_length` - it pulls the leading zero count straight from `value.class.zero` and uses this for the type's  bit size. If that gets baked into the `[U]Int` types at any point, this can be pointed straight to that, but should still be suitably efficient with the call straight to the Intrinsic.

Reply:
Of course! I completely forgot that `Int` wrapped non primitives too. There's a existing integer square root method in libgmp. I've added a binding for that and slapped it onto `BigInt#sqrt`. I've kept that one out of the `Math` module as there wasn't an existing dependency there on `big`, but can add that in if desired.